core/src/main/java/org/infinispan/interceptors/compat/TypeConverterStream.java

```diff
@@ -0,0 +1,66 @@
+package org.infinispan.interceptors.compat;
+
+import org.infinispan.CacheStream;
+import org.infinispan.commons.util.IteratorMapper;
+import org.infinispan.compat.TypeConverter;
+import org.infinispan.container.InternalEntryFactory;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.stream.impl.AbstractDelegatingCacheStream;
+import org.infinispan.stream.impl.spliterators.IteratorAsSpliterator;
+
+import java.util.Iterator;
+import java.util.Spliterator;
+
+/**
+ * Delegating stream that converts variables.  Note this stream specifically doesn't specify generics as it can have
+ * intermediate operations performed upon it that will change the type without changing the instance.  Also depending
+ * on if CacheEntry values are returned we have to unbox those properly.
+ */
+public class TypeConverterStream extends AbstractDelegatingCacheStream {
+   public TypeConverterStream(CacheStream<?> stream, TypeConverter<Object, Object, Object, Object> converter,
+           InternalEntryFactory entryFactory) {
+      super(stream);
+      this.converter = converter;
+      this.entryFactory = entryFactory;
+   }
+
+   private final TypeConverter<Object, Object, Object, Object> converter;
+   private final InternalEntryFactory entryFactory;
+
+   @Override
+   public Iterator<Object> iterator() {
+      // Note that a transformation could have been applied intermediately which means this
+      // may not be a cache entry instance at all - so we have to define it as Object first
+      return new IteratorMapper<>(super.iterator(), (java.lang.Object e) -> {
+         if (e instanceof CacheEntry) {
+            return convert((CacheEntry<?, ?>) e, converter, entryFactory);
+         } else {
+            // If it isn't a CacheEntry just unbox as is
+            // TODO: should we worry about arrays?
```

Comment:
The more I think about this should be fine.  And tbh it isn't even possible since the only intermediate operations we allow for remote invocation based streams is using our keyfilter converter framework which we control.  I can just remove this.